units/log_output_tests.cc

```diff
@@ -0,0 +1,156 @@
+// each `LogLine()` call, it doens't need to aquire a mutex.
+class FileOutput : public LogOutput {
+ public:
+  FileOutput(const char *filename) {
+    if (!filename) {
+      out_ = stderr;
+      return;
+    }
+
+    // Since multiple layers open the same file and write to it at the same
+    // time, it's opened in the append mode.
+    out_ = fopen(filename, "a");
+    if (!out_) {
+      SPL_LOG(ERROR) << "Failed to open " << filename
+                     << ". Using stderr as the alternative output.";
+      out_ = stderr;
+    }
+  }
+
+  ~FileOutput() {
+    if (out_ && out_ != stderr) {
+      fclose(out_);
+    }
+  }
+
+  void Flush() override {
+    assert(out_);
+    fflush(out_);
+  }
+
+  void LogLine(const std::string &line) override {
+    assert(out_);
+    fprintf(out_, "%s\n", line.c_str());
+    Flush();
+  }
+
+ private:
+  FILE *out_ = nullptr;
+};
+
+// This class is used for testing. It writes the data to a string
+// instead of a file. The data can be read using the `GetLog()` method.
+class StringOutput : public LogOutput {
+ public:
+  StringOutput(){};
+
+  void Flush() override {}
+
+  void LogLine(const std::string &line) override { out_.push_back(line); }
+
+  const std::vector<std::string> &GetLog() { return out_; }
+
+ private:
+  std::vector<std::string> out_ = {};
+};
+
+TEST(LogOutput, SingleThreadedLog) {
+  const char *filename = "test.log";
+  const std::string line = "name:event_name,timestamp:1234";
+  FileOutput file_out(filename);
+  file_out.LogLine(line);
+
+  std::ifstream test_file;
+  test_file.open(filename);
+  EXPECT_TRUE(test_file.is_open());
+  std::string stored_line;
+  getline(test_file, stored_line);
+  EXPECT_EQ(stored_line, line);
+  test_file.close();
+  remove(filename);
+
+  StringOutput string_out;
+  string_out.LogLine(line);
+  EXPECT_THAT(string_out.GetLog(), ElementsAre(line));
+}
+
+// This functions is executed by multiple threads. It logs the input to the file
+// as many times as `kLogsPerThread`.
+void LogToFileOutput(FileOutput *file_out, const std::string &line) {
+  for (size_t i = 0; i < kLogsPerThread; ++i) file_out->LogLine(line);
+}
+
+TEST(LogOutput, MultiThreadedFileLog) {
+  const char *filename = "test.log";
+  FileOutput file_out(filename);
+  const std::string line = "name:event_name,timestamp:1234";
+  std::thread threads[kThreadCount];
+
+  for (size_t i = 0; i < kThreadCount; ++i)
+    threads[i] = std::thread(LogToFileOutput, &file_out, line);
+
+  for (size_t i = 0; i < kThreadCount; ++i) threads[i].join();
+  ;
+
+  std::ifstream test_file;
+  test_file.open(filename);
+  EXPECT_TRUE(test_file.is_open());
+  std::string stored_line;
+  for (size_t i = 0; i < kThreadCount * kLogsPerThread; ++i) {
+    getline(test_file, stored_line);
+    EXPECT_EQ(stored_line, line);
+  }
+  getline(test_file, stored_line);
+  EXPECT_EQ(stored_line, "");
+  test_file.close();
+  remove(filename);
```

Comment:
If we create this in a temp dir we should not have to remove the test files, so that someone can read the log file contents when they have to debug it